src/raps/actions/crosschainSwap.ts

```diff
@@ -0,0 +1,239 @@
+  };
+
+  logger.debug(
+    'FILLCROSSCHAINSWAP',
+    tradeDetails,
+    transactionParams,
+    walletToUse.address,
+    permit,
+    chainId
+  );
+  return fillCrosschainQuote(tradeDetails, transactionParams, walletToUse);
+};
+
+const crosschainSwap = async (
+  wallet: Wallet,
+  currentRap: Rap,
+  index: number,
+  parameters: RapExchangeActionParameters,
+  baseNonce?: number
+): Promise<number | undefined> => {
+  logger.log(`[${actionName}] base nonce`, baseNonce, 'index:', index);
+  const {
+    inputAmount,
+    tradeDetails,
+    permit,
+    chainId,
+    requiresApprove,
+  } = parameters as SwapActionParameters;
+  const { dispatch } = store;
+  const { accountAddress } = store.getState().settings;
+  const { inputCurrency, outputCurrency } = store.getState().swap;
+  const { gasFeeParamsBySpeed, selectedGasFee } = store.getState().gas;
+  const gasParams = parseGasParamsForTransaction(selectedGasFee);
+  // if swap isn't the last action, use fast gas or custom (whatever is faster)
+  const isL2 = isL2Network(
+    ethereumUtils.getNetworkFromChainId(parameters?.chainId || ChainId.mainnet)
+  );
+  const emptyGasFee = isL2
+    ? !gasParams.gasPrice
+    : !gasParams.maxFeePerGas || !gasParams.maxPriorityFeePerGas;
+
+  if (currentRap.actions.length - 1 > index || emptyGasFee) {
+    const fastMaxFeePerGas =
+      gasFeeParamsBySpeed?.[gasUtils.FAST]?.maxFeePerGas?.amount;
+    const fastMaxPriorityFeePerGas =
+      gasFeeParamsBySpeed?.[gasUtils.FAST]?.maxPriorityFeePerGas?.amount;
+
+    if (greaterThan(fastMaxFeePerGas, gasParams?.maxFeePerGas || 0)) {
+      gasParams.maxFeePerGas = fastMaxFeePerGas;
+    }
+    if (
+      greaterThan(
+        fastMaxPriorityFeePerGas,
+        gasParams?.maxPriorityFeePerGas || 0
+      )
+    ) {
+      gasParams.maxPriorityFeePerGas = fastMaxPriorityFeePerGas;
+    }
+  }
+  let gasLimit;
+  try {
+    const newGasLimit = await estimateCrosschainSwapGasLimit({
+      chainId: Number(chainId),
+      requiresApprove,
+      tradeDetails: tradeDetails as CrosschainQuote,
+    });
+    gasLimit = newGasLimit;
+  } catch (e) {
+    logger.sentry(`[${actionName}] error estimateSwapGasLimit`);
+    captureException(e);
+    throw e;
+  }
+
+  let swap;
+  try {
+    logger.sentry(`[${actionName}] executing rap`, {
+      ...gasParams,
+      gasLimit,
+    });
+    const nonce = baseNonce ? baseNonce + index : undefined;
+
+    const swapParams = {
+      ...gasParams,
+      chainId,
+      flashbots: !!parameters.flashbots,
+      gasLimit,
+      nonce,
+      permit: !!permit,
+      tradeDetails,
+      wallet,
+    };
+
+    // @ts-ignore
+    swap = await executeCrosschainSwap(swapParams);
+    dispatch(
+      additionalDataUpdateL2AssetToWatch({
+        hash: swap?.hash || '',
+        inputCurrency,
+        network: ethereumUtils.getNetworkFromChainId(Number(chainId)),
+        outputCurrency,
+        userAddress: accountAddress,
+      })
+    );
```

Comment:
```suggestion
    if (swap?.hash) {
      dispatch(
        additionalDataUpdateL2AssetToWatch({
          hash: swap.hash,
          inputCurrency,
          network: ethereumUtils.getNetworkFromChainId(Number(chainId)),
          outputCurrency,
          userAddress: accountAddress,
        })
      );
    }
```

I realize this is just a copy-paste (I wrote this block originally I think), but I'm realizing invoking this code without a hash is useless and might lead to bad behavior. Seems better to just wrap in a conditional.

Reply:
👌